src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs

```diff
@@ -0,0 +1,65 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Reflection.Metadata
+{
+    public static class MetadataUpdater
+    {
+        /// <summary>
+        /// Updates the specified assembly using the provided metadata, IL and PDB deltas.
+        /// </summary>
+        /// <remarks>
+        /// Currently executing methods will continue to use the existing IL. New executions of modified methods will
+        /// use the new IL. Different runtimes may have different limitations on what kinds of changes are supported,
+        /// and runtimes make no guarantees as to the state of the assembly and process if the delta includes
+        /// unsupported changes.
+        /// </remarks>
+        /// <param name="assembly">The assembly to update.</param>
+        /// <param name="metadataDelta">The metadata changes to be applied.</param>
+        /// <param name="ilDelta">The IL changes to be applied.</param>
+        /// <param name="pdbDelta">The PDB changes to be applied.</param>
+        /// <exception cref="ArgumentNullException">The assembly argument is null.</exception>
+        /// <exception cref="NotSupportedException">The update could not be applied.</exception>
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta)
+        {
+            if (assembly is not RuntimeAssembly runtimeAssembly)
+            {
+                if (assembly is null) throw new ArgumentNullException(nameof(assembly));
+                throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);
+            }
+
+            // System.Private.CoreLib is not editable
+            if (runtimeAssembly == typeof(AssemblyExtensions).Assembly)
+                throw new InvalidOperationException (SR.InvalidOperation_AssemblyNotEditable);
+
+            unsafe
+            {
+                IntPtr monoAssembly = runtimeAssembly.GetUnderlyingNativeHandle ();
+                fixed (byte* metadataDeltaPtr = metadataDelta, ilDeltaPtr = ilDelta, pdbDeltaPtr = pdbDelta)
+                {
+                    ApplyUpdate_internal(monoAssembly, metadataDeltaPtr, metadataDelta.Length, ilDeltaPtr, ilDelta.Length, pdbDeltaPtr, pdbDelta.Length);
+                }
+            }
+        }
+
+        private static Lazy<string> s_ApplyUpdateCapabilities = new Lazy<string>(() => InitializeApplyUpdateCapabilities());
+
+        public static string GetCapabilities() => s_ApplyUpdateCapabilities.Value;
+
+        private static string InitializeApplyUpdateCapabilities()
+        {
+            return ApplyUpdateEnabled() != 0 ? "Baseline" : string.Empty ;
+        }
+
+        public static bool IsSupported => Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "";
```

Comment:
@lambdageek we want to introduce a feature switch backed up by msbuild property to use it to control build dependencies, right?

Reply:
Yea, we need a linker substitution for this flag, and also a runtime check

Reply:
```suggestion
        public static bool IsSupported => (Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "") && ApplyUpdateEnabled();
```

Reply:
We may need a new icall like ApplyUpdateEnabled but adds check if the other conditions for IsSupported (see the above thread with Jan about the coreclr version).

Reply:
@mikem8361 I can fix it up in a follow-up PR, or if you want to take a stab at it, I think you can just modify the implementation of ApplyUpdateEnabled 

https://github.com/dotnet/runtime/blob/0416c3469b7bccff087b89db3d9967dfd36070a3/src/mono/mono/metadata/icall.c#L5790

To something like

```c
return mono_metadata_update_available () && mono_metadata_update_enabled (NULL);
```

(The other callsite of `ApplyUpdateEnabled` is the `GetCapabilities` function - and I think the new version of ApplyUpdateEnabled makes sense there too - we will only return non-empty capabilities if the current running runtime is in a position to apply updates)

Reply:
Done